src/vos/vos_ilog.c

```diff
@@ -420,8 +444,17 @@ vos_ilog_punch_(struct vos_container *cont, struct ilog_df *ilog,
 		return rc;
 	}
 
+	if (dth) {
+		minor_epc = dth->dth_op_seq;
+	} else if (replay) {
+		/* If it's a replay, punch lower than the max in case there
+		 * are later visible updates with same major epoch.
+		 */
+		minor_epc--;
```

Comment:
I don't qutie get this. What's the differece of VOS_MINOR_EPC_MAX and VOS_MINOR_EPC_MAX - 1 in rebuild case? Any special minor epoch will be used for rebuild update?

Reply:
The issue is that rebuild needs to replay punches and in the case where there is a subsequent update at the same major epoch, it needs to punch everything else but not conflict with the replayed update that covers the punch.   Update will use MAX where punch will use MAX - 1

Reply:
I see, thanks. Then this "minor_epc--" for rebuild looks unnecessary to me, because vos_epc_punched() only return 'punched' when "punch_minor_epc > update_minor_epc", did I miss anything?

Reply:
Rebuild uses the next_punch field.  When iterating through a visible range, we return the nearest future punch epoch

169                 if (entry->ie_id.id_epoch > epoch) {
170                         if (ilog_has_punch(entry) &&
171                             entry->ie_status == ILOG_COMMITTED)
172                                 info->ii_next_punch = entry->ie_id.id_epoch;
173                         continue;
174                 }

This gets set if the entry has a punch (non-zero punch minor epoch) even if it's covered by an update.